src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,217 @@ public static CFProxySettings GetSystemProxySettings ()
 			dict.DangerousRelease ();
 			return new CFProxySettings (dict);
 		}
-		
-#if notyet
-		// FIXME: These require bindings for CFRunLoopSource and CFStreamClientContext
-		
-		public delegate void CFProxyAutoConfigurationResultCallback (IntPtr client, NSArray proxyList, NSError error);
-		
+
+		delegate void CFProxyAutoConfigurationResultCallbackInternal (IntPtr client, IntPtr proxyList, IntPtr error);
+		public delegate void CFProxyAutoConfigurationResultCallback (NSObject client, CFProxy[] proxyList, NSError error);
+		// helper delegate to reuse code
+		delegate IntPtr CreatePACCFRunLoopSource (CFProxyAutoConfigurationResultCallbackInternal cb, ref CFStreamClientContext context);
+
+		static CFProxy[] ParseProxies (IntPtr proxyList)
+		{
+			CFProxy[] proxies = null;
+			if (proxyList != IntPtr.Zero) {
+				using (var array = new CFArray (proxyList, false)) {
+					proxies = new CFProxy [array.Count];
+					for (int i = 0; i < proxies.Length; i++) {
+						var dict = new NSDictionary (array.GetValue (i));
+						proxies[i] = new CFProxy (dict);
+					}
+				}
+			}
+			return proxies;
+		}
+
+		static async Task<(CFProxy[] proxies, NSError error)> ExecutePacCFRunLoopSourceAsync (CreatePACCFRunLoopSource factory, CancellationToken cancellationToken)
+		{
+			CFProxy[] proxies = null;
+			NSError outError = null;
+			await Task.Run (() => {
+				// we need the runloop of THIS thread, so it is important to get it in the correct context
+				var runLoop = CFRunLoop.Current;
+				CFProxyAutoConfigurationResultCallbackInternal cb = delegate (IntPtr client, IntPtr proxyList, IntPtr error) {
+					if (error != IntPtr.Zero)
+						outError = new NSError (error);
+					else
+						proxies = ParseProxies (proxyList);
+					runLoop.Stop ();
+				};
+				var clientContext = new CFStreamClientContext ();
+				using (var loopSource = new CFRunLoopSource (factory (cb, ref clientContext)))
+				using (var mode = new NSString ("Xamarin.iOS.Proxy")) {
+					runLoop.AddSource (loopSource, mode);
+					// blocks until stop is called, will be done in the cb set previously
+					runLoop.RunInMode (mode, double.MaxValue, false);
+					runLoop.RemoveSource (loopSource, mode);
+				}
+			}, cancellationToken).ConfigureAwait (false);
```

Comment:
`await` spawns a new thread that asynchronously runs some code...what does `.ConfigureAwait(false)` (or `true`) do?

Reply:
Await spawns a thread yes does it always not block the thread, false.. You can imaging the await keyboard as to say 'execute this piece of code **somewhere** (we will get to the somewhere later) and get me back here to the same thread. How does .Net bring you back to the same thread? To do that, there is something called the SynchronizationContext. You can imaging everything as follows:

The following is the code you write:
```csharp
await GetMyDataAsync();
RestOfMethod();
```
This is the code *'generated'* by the runtime (not exactly, but more or less)
```csharp
var t = GetMyDataAsync();
var currentContext = SynchronizationContext.Current;
t.ContinueWith(delegate
{
    if (currentContext == null)
        RestOfMethod();
    else
        currentContext.Post(delegate { RestOfMethod(); }, null);
}, TaskScheduler.Current);
```
The above means, that when you are using the current synchronisation context, the task will be ran in it, which means that, if you are in the main thread (UI thread) the call **WILL BLOCK** the thread.

In our case, we do not want to be executed in the same sync context, we want it to be ran 'anywhere else' so we tell the system that we do not want to use the same sync context as the current one via **ConfigureAwait (false)**.

This is the most common use if you are writing libs or you do not want to block the current thread by using the same sync context (sometimes you do, since there is certain information we need in the current context etc.. (

Reply:
Thank you for the detailed explanation - that makes a lot of sense! :)